src/DIRAC/FrameworkSystem/Agent/MonitoringAgent.py

```diff
@@ -0,0 +1,544 @@
+    self.log.info("%s's process with ID: %s has been terminated successfully" % (agentName, process.pid))
+
+  def execute(self):
+    """Execute checks for agents, executors, services."""
+    for instanceType in ('executor', 'agent', 'service'):
+      for name, options in getattr(self, instanceType + 's').iteritems():
+        # call checkAgent, checkExecutor, checkService
+        res = getattr(self, 'check' + instanceType.capitalize())(name, options)
+        if not res['OK']:
+          self.logError('Failure when checking %s' % instanceType, '%s, %s' % (name, res['Message']))
+
+    res = self.componentControl()
+    if not res['OK']:
+      if 'Stopped does not exist' not in res['Message'] and \
+         'Running does not exist' not in res['Message']:
+        self.logError('Failure to control components', res['Message'])
+
+    if not self.errors:
+      res = self.checkURLs()
+      if not res['OK']:
+        self.logError('Failure to check URLs', res['Message'])
+    else:
+      self.logError('Something was wrong before, not checking URLs this time')
+
+    self.sendNotification()
+
+    if self.errors:
+      return S_ERROR('Error during this cycle, check log')
+
+    return S_OK()
+
+  @staticmethod
+  def getLastAccessTime(logFileLocation):
+    """Return the age of log file."""
+    lastAccessTime = 0
+    try:
+      lastAccessTime = os.path.getmtime(logFileLocation)
+      lastAccessTime = datetime.fromtimestamp(lastAccessTime)
+    except OSError as e:
+      return S_ERROR('Failed to access logfile %s: %r' % (logFileLocation, e))
+
+    now = datetime.now()
+    age = now - lastAccessTime
+    return S_OK(age)
+
+  def restartInstance(self, pid, instanceName, enabled):
+    """Kill a process which is then restarted automatically."""
+    if not (self.enabled and enabled):
+      self.log.info('Restarting is disabled, please restart %s manually' % instanceName)
+      self.accounting[instanceName]['Treatment'] = 'Please restart it manually'
+      return S_OK(NO_RESTART)
+
+    if any(pattern in instanceName for pattern in self.doNotRestartInstancePattern):
+      self.log.info('Restarting for %s is disabled, please restart it manually' % instanceName)
+      self.accounting[instanceName]['Treatment'] = 'Please restart it manually'
+      return S_OK(NO_RESTART)
+
+    try:
+      agentProc = psutil.Process(int(pid))
+      processesToTerminate = agentProc.children(recursive=True)
+      processesToTerminate.append(agentProc)
+
+      for proc in processesToTerminate:
+        proc.terminate()
+
+      _gone, alive = psutil.wait_procs(processesToTerminate, timeout=5,
+                                       callback=partial(self.on_terminate, instanceName))
+      for proc in alive:
+        self.log.info('Forcefully killing process %s' % proc.pid)
+        proc.kill()
+
+      return S_OK()
+
+    except psutil.Error as err:
+      self.logError('Exception occurred in terminating processes', '%s' % err)
```

Comment:
Should this using `log.exception` so the traceback is included as well?

Reply:
I don't think it helps any to know where inside psutil this comes from.
The message looks like
```
Errors:Exception occurred in terminating processes psutil.NoSuchProcess no process found with pid 26024
```
which probably means the service has an issue starting up and gets  a new PID between the ping failure and our attempt to kill it.